### PR TITLE
JIRA-1158

### PR DESCRIPTION
**Pull request by yukselakinci**
Giraph changes to support time estimation

### CODE REVIEW

File: giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java [hunk 1]
```diff
@@ -129,6 +135,36 @@ public void run() {
...
 
   /**
+   * Determine if the job will finish in allotted time.
+   * @param conf Giraph configuration
+   * @param job Job
+   * @param progress Current superstep progress.
+   * @return true it the job can finish in allotted time, false otherwise.
+   */
+  protected boolean canFinishInTime(GiraphConfiguration conf, Job job,
+      SuperstepProgress progress) {
```
Comment (majakabiljo) [inline giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java]: Why not just give combinedWorkerProgress to this method?

File: giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java [hunk 2]
```diff
@@ -129,6 +135,36 @@ public void run() {
...
+    try {
+      LOG.error("Killing the job because it won't complete in max " +
+              "allotted time: " +
+              GiraphConstants.MAX_ALLOWED_JOB_TIME_MS.get(conf) / 1000 + "s");
+      job.killJob();
+    } catch (IOException e) {
+      LOG.error(
+        "Failed to kill the job which won't complete in max allotted time: ",
+        e);
+    }
```
Comment (majakabiljo) [inline giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java]: Please rebase and use killJobWithMessage method

File: giraph-core/src/main/java/org/apache/giraph/job/CombinedWorkerProgress.java [hunk 3]
```diff
@@ -144,6 +143,38 @@ public CombinedWorkerProgress(Iterable<WorkerProgress> workerProgresses,
...
+   */
+  public long getCurrentSuperstep() {
+    return currentSuperstep;
+  }
+
+  /**
+   * Get workers in superstep
+   * @return Workers in superstep.
+   */
+  public long getWorkerseInSuperstep() {
```
Comment (majakabiljo) [inline giraph-core/src/main/java/org/apache/giraph/job/CombinedWorkerProgress.java]: Typo

File: giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java [hunk 4]
```diff
@@ -129,6 +137,20 @@ public void run() {
   }
 
   /**
+   * Determine if the job will finish in allotted time
+   * @param conf Giraph configuration
+   * @param job Job
+   * @param workerCount Worker count
```
Comment (majakabiljo) [inline giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java]: Do you need workerCount here, should you use workersInSuperstep? If you need it, rename to startedWorkerCount or something, this way it sounds like how many workers job needs.

**Review by majakabiljo — COMMENTED** [review]
Thanks, final comments

File: giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java (hunk 4 above)
Comment (yukselakinci) [reply giraph-core/src/main/java/org/apache/giraph/job/DefaultJobProgressTrackerService.java]: this is not the started worker count, but all the worker counts that are requested by the job. My implementation in Digraph will only calculate the expected completion time if all workers are included in the CombinedWorkerCount to make the calculation accurate .

File: giraph-core/src/main/java/org/apache/giraph/job/CombinedWorkerProgress.java (hunk 3 above)
Comment (yukselakinci) [reply giraph-core/src/main/java/org/apache/giraph/job/CombinedWorkerProgress.java]: Thanks

**Review by majakabiljo — APPROVED** [review]
Lgtm, I'll commit it

**Pull request closed without merging**